File: apps/desktop/src/lib/tauri-commands/whats-new.ts
// "What's new" popup: fetch the changelog slice and read the dev simulate-update override.

import { commands } from '$lib/ipc/bindings'
import type { WhatsNewRelease, WhatsNewSection } from '$lib/ipc/bindings'

export type { WhatsNewRelease, WhatsNewSection }

/**
 * Returns the displayable releases in `sinceVersion < release <= current`, newest first,
 * truncated to `max`. `sinceVersion = null` means no lower bound (the latest `max`).
 *
 * The result can be empty even on a real "show" decision (every in-range release dropped
 * to nothing backend-side), so the auto-popup caller collapses an empty result to a stamp.
 */
export function getWhatsNew(sinceVersion: string | null, max: number): Promise<WhatsNewRelease[]> {
  return commands.getWhatsNew(sinceVersion, max)
}

/**
 * Reads the `CMDR_SIMULATE_UPDATE_FROM` dev flag (a backend-process env var invisible to
 * the Vite frontend). Returns the simulated "updated from" version, or `null` when unset.
 */
export function whatsNewDevOverride(): Promise<string | null> {
  return commands.whatsNewDevOverride()
}

File: apps/desktop/src-tauri/src/commands/whats_new.rs
//! IPC commands for the "What's new" popup.
//!
//! Thin pass-throughs over the `whats_new` module. The changelog is embedded in
//! the binary (`include_str!`), so these commands touch no filesystem at runtime
//! and need no `blocking_with_timeout`: there's nothing to hang on.

use crate::whats_new::{self, WhatsNewRelease};

/// Returns the displayable releases in `since_version < v <= current`, newest
/// first, truncated to `max`. `since_version = None` means no lower bound (the
/// latest `max`). "Current" is the running binary's version.
#[tauri::command]
#[specta::specta]
pub fn get_whats_new(since_version: Option<String>, max: u32) -> Vec<WhatsNewRelease> {
    let current = env!("CARGO_PKG_VERSION");
    whats_new::releases_between(since_version.as_deref(), current, max as usize)
}

/// Surfaces the `CMDR_SIMULATE_UPDATE_FROM` dev flag to the frontend: when set,
/// returns the version string to diff from so a dev session can force the startup
/// popup without hand-editing `settings.json`. The env var is a backend-process
/// value the Vite frontend can't read directly, so it crosses through here.
/// Mirrors the `CMDR_MOCK_LICENSE` / `CMDR_MOCK_FDA` dev-flag family.
#[tauri::command]
#[specta::specta]
pub fn whats_new_dev_override() -> Option<String> {
    std::env::var("CMDR_SIMULATE_UPDATE_FROM")
        .ok()
        .filter(|v| !v.is_empty())
}
